Add ICE failing tests

CohenArthur opened the pull request:
As per @philberty 's advice, I've tried finding a few failing test cases for the compiler. In both of those cases, the behaviour is unexpected: `match_self.rs` produces a segfault, while `cfg_feature.rs` ends up in an infinite loop. Would you like me to also open up issues so that they can be documented?

I'd also love to start hacking at the compiler to try and figure them out 

Here is the log for both files:

<details><summary>match_self.rs</summary>
<p>

```sh
>  gccrs ../gccrs/gcc/testsuite/rust.test/compilable/match_self.rs 
Preparing to parse files.
Attempting to parse file: ../gccrs/gcc/testsuite/rust.test/compilable/match_self.rs
beginning null denotation self/self-alias/dollar/crate/super handling
current peek token when starting path pratt parse: '{'
current token (just about to return path to null denotation): '{'
just finished parsing path (going to disambiguate) - peeked token is '{'
about to start parsing match arm patterns
successfully parsed initial match arm pattern
successfully parsed match arm patterns
successfully parsed match arm
successfully parsed inherent impl
SUCCESSFULLY PARSED CRATE
ran register_plugins (with no body)
SUCCESSFULLY REGISTERED PLUGINS
started injection
finished injection
SUCCESSFULLY FINISHED INJECTION
started expansion
finished expansion
SUCCESSFULLY FINISHED EXPANSION
rust1: internal compiler error: Segmentation fault
0xe9170f crash_signal
	../../gccrs/gcc/toplev.c:330
0x98e2ea Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*, bool)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-expr.h:41
0x98e773 Rust::Resolver::TypeCheckStmt::visit(Rust::HIR::ExprStmtWithBlock&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-stmt.h:44
0x989e81 Rust::Resolver::TypeCheckStmt::Resolve(Rust::HIR::Stmt*, bool)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-stmt.h:38
0x989e81 operator()
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check.cc:93
0x989e81 __invoke_impl<bool, Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)::<lambda(Rust::HIR::Stmt*)>&, Rust::HIR::Stmt*>
	/usr/include/c++/10.2.0/bits/invoke.h:60
0x989e81 __invoke_r<bool, Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)::<lambda(Rust::HIR::Stmt*)>&, Rust::HIR::Stmt*>
	/usr/include/c++/10.2.0/bits/invoke.h:141
0x989e81 _M_invoke
	/usr/include/c++/10.2.0/bits/std_function.h:291
0x98a4f6 std::function<bool (Rust::HIR::Stmt*)>::operator()(Rust::HIR::Stmt*) const
	/usr/include/c++/10.2.0/bits/std_function.h:622
0x98a4f6 Rust::HIR::BlockExpr::iterate_stmts(std::function<bool (Rust::HIR::Stmt*)>)
	../../gccrs/gcc/rust/hir/tree/rust-hir-expr.h:2576
0x98a4f6 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check.cc:88
0x98e311 Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*, bool)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-expr.h:41
0x98f489 Rust::Resolver::TypeCheckImplItem::visit(Rust::HIR::Method&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-implitem.h:246
0x98c82d Rust::Resolver::TypeCheckImplItem::Resolve(Rust::HIR::InherentImplItem*, Rust::TyTy::BaseType*)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-implitem.h:187
0x98c82d Rust::Resolver::TypeCheckItem::visit(Rust::HIR::InherentImpl&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-item.h:55
0x98a29b Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item*)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-item.h:40
0x98a29b Rust::Resolver::TypeResolution::Resolve(Rust::HIR::Crate&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check.cc:42
0x8b9661 Rust::Session::parse_file(char const*)
	../../gccrs/gcc/rust/rust-session-manager.cc:544
0x8b9a7e Rust::Session::parse_files(int, char const**)
	../../gccrs/gcc/rust/rust-session-manager.cc:434
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
```

</p>
</details>

<details><summary>cfg_feature.rs</summary>
<p>

```sh
checked key-value pair for cfg: 'feature', 'somefeature' - is not in target data
asked to check cfg of attrinputmetaitemcontainer - delegating to first item. container: '(feature = "somefeature", attribute = "someattr")'
checked key-value pair for cfg: 'feature', 'somefeature' - is not in target data
asked to check cfg of attrinputmetaitemcontainer - delegating to first item. container: '(feature = "somefeature", attribute = "someattr")'
checked key-value pair for cfg: 'feature', 'somefeature' - is not in target data
asked to check cfg of attrinputmetaitemcontainer - delegating to first item. container: '(feature = "somefeature", attribute = "someattr")'
checked key-value pair for cfg: 'feature', 'somefeature' - is not in target data
asked to check cfg of attrinputmetaitemcontainer - delegating to first item. container: '(feature = "somefeature", attribute = "someattr")'
checked key-value pair for cfg: 'feature', 'somefeature' - is not in target data
asked to check cfg of attrinputmetaitemcontainer - delegating to first item. container: '(feature = "somefeature", attribute = "someattr")'
checked key-value pair for cfg: 'feature', 'somefeature' - is not in target data
asked to check cfg of attrinputmetaitemcontainer - delegating to first item. container: '(feature = "somefeature", attribute = "someattr")'
checked key-value pair for cfg: 'feature', 'somefeature' - is not in target data
asked to check cfg of attrinputmetaitemcontainer - delegating to first item. container: '(feature = "somefeature", attribute = "someattr")'
checked key-value pair for cfg: 'feature', 'somefeature' - is not in target data
asked to check cfg of attrinputmetaitemcontainer - delegating to first item. container: '(feature = "somefeature", attribute = "someattr")'
checked key-value pair for cfg: 'feature', 'somefeature' - is not in target data
asked to check cfg of attrinputmetaitemcontainer - delegating to first item. container: '(feature = "somefeature", attribute = "someattr")'
checked key-value pair for cfg: 'feature', 'somefeature' - is not in target data
asked to check cfg of attrinputmetaitemcontainer - delegating to first item. container: '(feature = "somefeature", attribute = "someattr")'
checked key-value pair for cfg: 'feature', 'somefeature' - is not in target data
asked to check cfg of attrinputmetaitemcontainer - delegating to first item. container: '(feature = "somefeature", attribute = "someattr")'
checked key-value pair for cfg: 'feature', 'somefeature' - is not in target data
asked to check cfg of attrinputmetaitemcontainer - delegating to first item. container: '(feature = "somefeature", attribute = "someattr")'
...
```

</p>
</details>

philberty commented on gcc/testsuite/rust.test/compilable/match_self.rs:
```diff
@@ -0,0 +1,12 @@
+
+impl Struct {
+    pub fn method(&mut self) {
+        match self {
```
Can you create another test case similar to this but without the match something like:

fn method(&mut self) ->i32 {
  123
}

fn main() {
   let a:Struct;
  a.method()
}

Currently #241 is not implemented yet

CohenArthur replied:
Sure thing!

CohenArthur replied:
Without the `match`, the test compiles successfully. I added it in 70ab13c

philberty replied:
Interesting that it passes lol I guess GCC might be quite permissive but I bet if you run -fdump-tree-gimple and examine the gimple output that the generated method is receiving an instance and not a reference i would have it expected it to fail/crash lol thanks for adding this

philberty replied:
I will be tackling pattern matching closer to June time with the control flow 2 milestones that I have set out todo just for reference.

dkm commented:
or you could build using `--enable-checking=all` to see if somewhere some internal check does assert?

philberty commented:
> or you could build using `--enable-checking=all` to see if somewhere some internal check does assert?

i always for get about that option can you add that as an example to the README?

dkm commented:
> > or you could build using `--enable-checking=all` to see if somewhere some internal check does assert?
> 
> I always forget about that option can you add that as an example to the README?

Sure !

bjorn3 commented on gcc/testsuite/rust.test/compilable/cfg_feature.rs:
```diff
@@ -0,0 +1,7 @@
+use std::env;
+
+#[cfg_attr(feature = "somefeature", attribute = "someattr")]
```
Can you name is file `cfg_attr.rs`? `cfg` attributes don't have to be of the form `feature = "..."`. That is merely what Cargo uses. You can pass almost arbitrary values to `--cfg`. Eg `--cfg foo` will enable `#[cfg(foo)]` and `--cfg bar="baz"` will enable `#[cfg(bar = "baz")]`.

CohenArthur replied:
Done in b426501

philberty commented:
With the recent merge of the test suite update from @dkm in #301 can you add these test cases to rust.test/xfail_compile we will get an XPASS when things start to work for that case using the match expr.

CohenArthur commented:
By the way, considering that `cfg_attr.rs` produces an infinite-loop, is there anyway to add a timeout to the test-runner or to this test-case in particular? This would also remove a problem that I'm having when running the test suite now, which is that some of make's spawned processes are not killed when the `make check-rust` command is interrupted

bjorn3 commented:
Rust's compiletest merely gives a warning after 60s. Considering tests failed after a fixed amount of time would spuriously break them when they are near that time limit normally, it breaks running tests on slow systems and it breaks when the user uses ^Z to suspend the tests to for example attach a debugger or do something else that needs a lot of cpu time and then continues it after a while.

CohenArthur commented:
> Rust's compiletest merely gives a warning after 60s. Considering tests failed after a fixed amount of time would spuriously break them when they are near that time limit normally, it breaks running tests on slow systems and it breaks when the user uses ^Z to suspend the tests to for example attach a debugger or do something else that needs a lot of cpu time and then continues it after a while.

That's a good point, I hadn't thought about that. But then, merging this test would make it impossible to run the test suite anymore

dkm commented:
On March 28, 2021 10:33:49 AM UTC, CohenArthur ***@***.***> wrote:
>> Rust's compiletest merely gives a warning after 60s. Considering tests failed after a fixed amount of time would spuriously break them when they are near that time limit normally, it breaks running tests on slow systems and it breaks when the user uses ^Z to suspend the tests to for example attach a debugger or do something else that needs a lot of cpu time and then continues it after a while.
>
>That's a good point, I hadn't thought about that. But then, merging this test would make it impossible to run the test suite anymore
>

DejaGNU has some internal tool timeout, so it should still work, but that would force every run to reach it… I guess we should simply add the test along with the fix.

CohenArthur commented:
> I guess we should simply add the test along with the fix.

Done in #319

philberty commented:
bors try

CohenArthur commented:
Sorry about the lack of work on this, I'm currently extremely busy. I'll take care of the test issues really soon

tschwinge commented:
There is, by the way, a rather new DejaGnu directive in GCC that may be useful for you: `dg-ice`, to mark up testcases that are (currently) expected to ICE.  Read it up in `gcc/doc/sourcebuild.texi`.  Assuming that the ICE is in the Rust front end, the testcase should be put into `gcc/testsuite/rust.test/xfail_compile/`.

CohenArthur commented:
So thanks to @tschwinge 's advice, I've added the `dg-ice` directive to the test and moved it to `xfail_compile`, since the ICE happens in the Rust frontend. The tests pass on my machine. Thank you everyone for the advice

CohenArthur commented:
However the ICE does not seem to be caught by dejagnu. I might be using it wrong

```
Running /home/kagounard/Git/gccrs/gcc/testsuite/rust.test/xfail_compile/xfail_compile.exp ...
FAIL: rust.test/xfail_compile/match_self.rs (internal compiler error)
FAIL: rust.test/xfail_compile/match_self.rs (test for excess errors)

		=== rust Summary ===

# of expected passes		807
# of unexpected failures	2
# of expected failures		26
make[2]: Leaving directory '/home/kagounard/Git/gccrs-build/gcc'
make[1]: Leaving directory '/home/kagounard/Git/gccrs-build/gcc'
```

dkm commented:
> However the ICE does not seem to be caught by dejagnu. I might be using it wrong
> 
> ```
> Running /home/kagounard/Git/gccrs/gcc/testsuite/rust.test/xfail_compile/xfail_compile.exp ...
> FAIL: rust.test/xfail_compile/match_self.rs (internal compiler error)
> FAIL: rust.test/xfail_compile/match_self.rs (test for excess errors)
> 
> 		=== rust Summary ===
> 
> # of expected passes		807
> # of unexpected failures	2
> # of expected failures		26
> make[2]: Leaving directory '/home/kagounard/Git/gccrs-build/gcc'
> make[1]: Leaving directory '/home/kagounard/Git/gccrs-build/gcc'
> ```
> 
> (Despite match_self.rs being marked with `// dg-ice`)

Maybe we need to tune a bit more the output pruning to let the internal compiler error message be handled by dg... You can try to mimic what @tschwinge did with the error/warning and see if it works better ?

dkm commented:
See #384 for possible usage. HTH !

CohenArthur commented:
> See #384 for possible usage. HTH !

Thanks :D I tried looking at the dejagnu sources but couldn't figure out how to make `proc dg-ice` understand our output better. Turns out I was completely in the wrong place

philberty commented:
now that #384 is in do you want to try this again @CohenArthur

philberty commented:
bors try

CohenArthur commented:
So, funny thing @philberty: The tests in question do not ICE anymore :D I think this PR can be closed. Recent commits have also fixed the segmentation faults as a side effect. The main issue was
